### DISCUSSION TITLE
How do I change the model?

### DISCUSSION THREAD

**Discussion by bemorecyber-rab** [danielmiessler/Fabric#138] [Q&A] [answered]
Hi all,

Love this and want to get using it. I am using a Windows machine and got fabric installed using WSL + VS Code.

Went to use it and it says I need to use GPT-4 ... which is fine, I don't pay for it (yet!) but wanted to test this with the older models.
I saw the fabric -m MODEL command but when I run it, it just hangs.. or I have to fail it.

Has anyone else downgraded the model being used?

I can list models fine **"fabric --listmodels"** but none of them seems to work.

**Comment by ben0815** [comment, marked as answer]
pbpaste|fabric -m gpt-3.5-turbo -p extract_wisdom
works for me. It's the exact name from the listmodels command.

**Reply by bemorecyber-rab** [reply]
ahhh you are calling the model with the prompt ... I thought I could just change the model as the default - this makes sense

So I need to call the model and then call the pattern - makes sense. Thanks @ben0815

**Comment by nic0711** [comment]
It would be great if we could also use local Mistral-7b...  ;)

**Reply by ptsiampas** [reply]
I haven't looked at the config code in detail yet, but I am thinking of modifying which API it connects to so I can get Fabric to use a local LLM when I give it a switch.  I am not a great coder, but I am going to look it over this weekend to see if GPT4 can help me modify it ;P

**Reply by waldo323** [reply]
To help you get started and so you don't have to duplicate work:
https://github.com/danielmiessler/fabric/issues/101

https://knasmueller.net/running-fabric-locally-with-ollama

I believe there is a more official way to use local models in the future using  LIteLLM, based on the comments, but in the meantime there are ways to using Mistral with ollama (pretty sure it is the mostral7b I tried it with as well as a couple other models).  
There have been some updates to that utils.py file since the article was written but the change suggested in the article still works with the updated file.
Note that my copy and paste may have altered spacing so double check everything makes sense before trying

```
diff --git a/installer/client/cli/utils.py b/installer/client/cli/utils.py   
index 51896f0..025d2e0 100644                                                
--- a/installer/client/cli/utils.py                                          
+++ b/installer/client/cli/utils.py                                          
@@ -37,9 +37,14 @@ class 
Standalone:                                                  
env_file = 
os.path.expanduser(env_file)                                      
load_dotenv(env_file)                                                        
try:                                                                
-            apikey = os.environ["OPENAI_API_KEY"]                           
-            self.client = OpenAI()                                          
-            self.client.api_key = apikey                                    
+            #apikey = os.environ["OPENAI_API_KEY"]                          
+            #self.client = OpenAI()                                         
+            #self.client.api_key = apikey                                   
+            self.client = OpenAI(                                           
+                    base_url = 'http://localhost:11434/v1',
+                    api_key = 'ollama', #required but unused
+            )
+
         except KeyError:
             print("OPENAI_API_KEY not found in environment variables.")

```

**Reply by ptsiampas** [reply]
Thank you :)

**Reply by xntrik** [reply]
Even better, you can just set your env vars with OPENAI_BASE_URL=http://localhost:11434/v1

This gets picked up by the OpenAI Py library, and will "just work".

For example, if I edit my ~/.config/fabric/.env to look like:

```
OPENAI_API_KEY=llama
OPENAI_BASE_URL=http://localhost:11434/v1
```

Then this will work.

The `--listmodels` command won't work though, and I don't believe this supports streaming either.

**Reply by nic0711** [reply]
Yeah!
It works with LM Studio too...

```
OPENAI_API_KEY=lmstudio
OPENAI_BASE_URL=http://localhost:1234/v1
```
Of course you have to adjust the local port. 

Start LM Studio, choose your model, start the server - have fun.

**Reply by zestysoft** [reply]
Seems like I'm missing a step somewhere?

```
$ cat ~/.config/fabric/.env 
OPENAI_API_KEY=lmstudio
OPENAI_BASE_URL=http://localhost:8000/v1

$ echo "An idea that coding is like speaking with rules." | fabric --model mistral:instruct --pattern write_essay
Failed to fetch models: HTTP 401
```

EDIT:

I had to give it a real openai api key and re-run `fabric --setup`.  Afterwards, I had to install ollama, and run a model in order for `fabric --listmodels` to show a local model.  Finally I could supply the OPENAI_BASE_URL setting above and have it properly access the locally hosted model making sure to supply the model in the command line.

**Reply by nic0711** [reply]
For me it works without an OpenAI key - just start the local server with local model. 

Code 401: "Unauthorized"

**Reply by zestysoft** [reply]
Strange.  What happens if you run `fabric --listmodels`?

If I have a valid OPENAI_API_KEY, I get this output:
```
GPT Models:
gpt-3.5-turbo
gpt-3.5-turbo-0125
gpt-3.5-turbo-0301
gpt-3.5-turbo-0613
gpt-3.5-turbo-1106
gpt-3.5-turbo-16k
gpt-3.5-turbo-16k-0613
gpt-3.5-turbo-instruct
gpt-3.5-turbo-instruct-0914
gpt-4
gpt-4-0125-preview
gpt-4-0613
gpt-4-1106-preview
gpt-4-turbo-preview
gpt-4-vision-preview

Local Models:
mixtr

Claude Models:
claude-3-opus-20240229
```

However the moment I change the OPENAI_API_KEY value to something else like 'lmstudio', I again get that 'Failed to fetch models: HTTP 401' error.

Any chance you have that environment variable being set elsewhere?  Like from the terminal if you run `export|grep OPENAI` nothing shows up?

EDIT

It's hard coded here: https://github.com/danielmiessler/fabric/blob/24e161686489935a68891c269912fd5d8581b035/installer/client/cli/utils.py#L260

Fix here: https://github.com/danielmiessler/fabric/pull/189

**Comment by xntrik** [comment]
I've got a small PR to allow the default model to be set outside of cli flags: https://github.com/danielmiessler/fabric/pull/156

**Reply by xssdoctor** [reply]
It’s already added. —changeDefaultModel.

**Reply by xntrik** [reply]
Yass! Thank you @xssdoctor

**Comment by mgavish** [comment]
does this work for you?
`yt --transcript https://youtube.com/watch?v=uXs-zPc63kM | fabric --stream --pattern extract_wisdom`
